### PR TITLE
Deadchat Plays SS13: Ghost Control of Mobs

### PR DESCRIPTION
**Pull request by lewcc**
## What Does This PR Do
Ported from https://github.com/tgstation/tgstation/pull/55314.

Allows admins to give deadchat control over just about anything (immovable rods, singularities, Ian, your local clown) in classic Twitch-Plays style.
Control can be set in anarchy or democracy mode, with anarchy letting every move count as soon as possible (though putting users on individual cooldowns), and democracy picking the most popular option after a cooldown.

To answer your (likely) questions:
- You can only control the thing when you're orbiting it as a ghost.
- Your control messages *will not go to dchat*, and other people won't see them.
- Likewise, other messages about controlled objects have made will only be relayed to orbiters.
- Admins can start it by **a new option to grant deadchat control in the VV dropdown**, or by proc-calling `deadchat_plays`, with arguments for the mode and cooldown.
- this control can be ended similarly, a VV dropdown option to end control will appear on anything that's being dchat controlled.

## Why It's Good For The Game
It's fun! I think it's a fun way to appease deadchat for a little by giving them something productive to do instead of salting. Also, it could make some events a bit more interesting, if you ask me. I've added some support for humans, so punpun could be let loose on the station under deadchat's control, for example. This could also be expanded down the line into a midround or something.

## Images of changes
Controlling a human:
https://user-images.githubusercontent.com/89928798/223902032-c40d7f7d-e3e0-404a-bd02-c812caf93b69.mp4

Controlling a rod:
https://user-images.githubusercontent.com/89928798/223902104-8347f150-a58a-4642-b9e5-ffb4acfcf385.mp4

## Testing
Lots of controlling mobs.

## Changelog
:cl:
add: Admins can now set items up to be controlled by deadchat!
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/friendly/dog.dm [hunk 1]
```diff
@@ -105,6 +105,48 @@
 	..(gibbed)
 	regenerate_icons()
 
+/mob/living/simple_animal/pet/dog/corgi/deadchat_plays(mode = ANARCHY_MODE, cooldown = 12 SECONDS)
+	. = AddComponent(/datum/component/deadchat_control/cardinal_movement, mode, list(
+		"speak" = CALLBACK(src, .proc/handle_automated_speech, TRUE),
+		"wear_hat" = CALLBACK(src, .proc/find_new_hat),
+		"drop_hat" = CALLBACK(src, .proc/drop_hat),
+		"spin" = CALLBACK(src, /mob.proc/emote, "spin")), cooldown, CALLBACK(src, PROC_REF(end_dchat_plays)))
```
[the remark this answers is not in the document]
Comment (lewcc) [reply code/modules/mob/living/simple_animal/friendly/dog.dm]: Fixed in https://github.com/ParadiseSS13/Paradise/pull/20600/commits/1a677564d60e0557b133341ae58e6dbb9a497f27

File: code/datums/components/deadchat_control.dm [hunk 2]
```diff
@@ -0,0 +1,243 @@
+///Will execute a single command after the cooldown based on player votes.
+#define DEMOCRACY_MODE (1<<0)
```
Comment (AffectedArc07) [inline code/datums/components/deadchat_control.dm:2]: Undef these please, or if you need them in other files, prefix them with ghostcontrol or something.

**Comment by lewcc** [comment]
https://user-images.githubusercontent.com/89928798/224117656-22e6728e-fd80-42b9-a23a-2c0451981fb1.mp4


Added a vv menu item to make admemery a bit easier (note I've cleaned it up since taking this video).

File: code/datums/components/deadchat_control.dm (hunk 2 above)
Comment (lewcc) [reply code/datums/components/deadchat_control.dm]: Addressed, moved em to out to a defines file as well

File: code/__DEFINES/misc_defines.dm [hunk 3]
```diff
@@ -538,3 +538,12 @@
...
 #define REFLECTABILITY_ENERGY 2
+
+// Deadchat control defines
+
+///Will execute a single command after the cooldown based on player votes.
+#define DEADCHAT_DEMOCRACY_MODE (1<<0)
+///Allows each player to do a single command every cooldown.
+#define DEADCHAT_ANARCHY_MODE (1<<1)
+///Mutes the democracy mode messages send to orbiters at the end of each cycle. Useful for when the cooldown is so low it'd get spammy.
+#define MUTE_DEMOCRACY_MESSAGES (1<<2)
```
Comment (AffectedArc07) [inline code/__DEFINES/misc_defines.dm:549]:
```suggestion
#define MUTE_DEATCHAT_DEMOCRACY_MESSAGES (1<<2)
```

File: code/datums/datumvars.dm [hunk 4]
```diff
@@ -1236,6 +1236,55 @@
 		var/datum/ui_module/colour_matrix_tester/CMT = new(target=target)
 		CMT.ui_interact(usr)
 
+	if(href_list["grantdeadchatcontrol"])
+		if(!check_rights(R_ADMIN))
+			return
```
Comment (AffectedArc07) [inline code/datums/datumvars.dm:1241]:
```suggestion
		if(!check_rights(R_EVENT))
			return
```

**Comment by lewcc** [comment]
> This might be overcomplicating it, but it would be cool to see a list of votecounts for the current vote cycle somewhere on screen maybe?

Maybe? I could see that being cool if done right. I think think that'd be something worth looking into once the base functionality is in, though.

**Comment by Probablyjay19** [comment]
Can't wait to create year long rivalries with fellow spessmen due to 6 people not deciding on wether or not we should eat a bar of soap as a Dchat controlled drask.

File: code/datums/components/deadchat_control.dm [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	if(QDELETED(parent) || !(deadchat_mode & DEADCHAT_DEMOCRACY_MODE))
+		deltimer(timerid)
+		return
+	var/result = count_democracy_votes()
+	if(!isnull(result))
+		inputs[result].Invoke()
+		if(!(deadchat_mode & MUTE_DEADCHAT_DEMOCRACY_MESSAGES))
+			var/message = "<span class='deadsay italics bold'>[parent] has done action [result]!<br>New vote started. It will end in [input_cooldown * 0.1] second\s.</span>"
+			for(var/mob/dead/observer/M in orbiters)
+				to_chat(M, message)
```
Comment (Sirryan2002) [inline code/datums/components/deadchat_control.dm:100]: will this send message to people with dsay disabled? If so, it needs to be able to be filtered due to players possibly not wanting their textbox spammed with a low-delay dchat plays ss13 game.

**Review by Sirryan2002 — CHANGES_REQUESTED** [review]
needs to be

File: code/modules/power/singularity/singularity.dm [hunk 6]
```diff
@@ -421,7 +421,7 @@
 		M.Stun(6 SECONDS)
 		M.visible_message("<span class='danger'>[M] stares blankly at [src]!</span>", \
 						"<span class='userdanger'>You look directly into [src] and feel weak.</span>")
-	return 
+	return
```
[the remark this answers is not in the document]
Comment (lewcc) [reply code/modules/power/singularity/singularity.dm:424]: This wasn't one of my (intended) changes, was just an automatic whitespace fixes. There are a few other weird quirks with the returns in this file, might be something to look at later.

[the remark this answers is not in the document]
Comment (Sirryan2002) [reply code/modules/power/singularity/singularity.dm:424]: you can just resolve this then

File: code/datums/components/deadchat_control.dm (hunk 5 above)
Comment (lewcc) [reply code/datums/components/deadchat_control.dm:100]: It should only send to people who are directly orbiting the object. That being said, I can definitely account for that on all these to_chats.

File: code/modules/power/singularity/singularity.dm (hunk 6 above)
[the remark this answers is not in the document]
Comment (lewcc) [reply code/modules/power/singularity/singularity.dm]: It actually got removed when merging master anyway, haha

File: code/datums/components/deadchat_control.dm (hunk 5 above)
Comment (lewcc) [reply code/datums/components/deadchat_control.dm:100]: Addressed, added some checks to account for deadchat mutes.

**Review by Sirryan2002 — APPROVED** [review]
honestly probably doesn't even need a TM since this is active only by admin intervention (Which means we can just not use it if it breaks). As long as this has been well-testing locally I'm fine with approving it

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**